### PR TITLE
Layout widget in editor is now scrollable

### PR DESCRIPTION
**Pull request by Semphriss**
Having dozens of tilemaps and other sector elements is no longer a pain to manage - now you can scroll the toolbar and access those sneaky little icons!

----------

> Urgh! I can't access all my tilemaps! They're going off-screen!

Screenshot from 2020-09-05 01-19-34

> Ah, much better.

Screenshot from 2020-09-05 01-19-47

### CODE REVIEW

**Comment by serano01** [comment]
Awesome, especially for the forest worldmap, since it has so many tilesets.

File: src/editor/layer_icon.cpp [hunk 1]
```diff
@@ -57,6 +57,25 @@ LayerIcon::draw(DrawingContext& context, const Vector& pos)
...
+LayerIcon::draw(DrawingContext& context, const Vector& pos, int pixels_shown)
+{
+  if (!is_valid()) return;
+
+  ObjectIcon::draw(context, pos, pixels_shown);
+  int l = get_zpos();
+  if (l != std::numeric_limits<int>::min()) {
+    //context.color().draw_text(Resources::small_font, std::to_string(l),
+    //                            pos + Vector(16,16),
+    //                            ALIGN_CENTER, LAYER_GUI, ColorScheme::Menu::default_color);
```
Comment (Zwatotem) [inline src/editor/layer_icon.cpp:70]: Looks like there's some leftover testing code here

File: src/editor/object_icon.cpp [hunk 2]
```diff
@@ -68,4 +68,16 @@ ObjectIcon::draw(DrawingContext& context, const Vector& pos)
...
 
+void
+ObjectIcon::draw(DrawingContext& context, const Vector& pos, int pixels_shown)
+{
+  auto cropped_surface = m_surface->region(Rect(32-pixels_shown,
+                                                0,
+                                                32,
+                                                32));
+  context.color().draw_surface_scaled(cropped_surface,
+                                      Rectf(pos + Vector(32-pixels_shown,0) + m_offset, pos + Vector(32,32) - m_offset),
```
Comment (Zwatotem) [inline src/editor/object_icon.cpp:79]: Operator spacing is inconsistent here and in some other places. Please adjust it to be always spaced (except commas of course).

**Review by Zwatotem — COMMENTED** [review]
Looks good, but I would block it from further scrolling if there's only one icon left on the screen. Currently you can flush them all to the left.
Btw, I expected mouse scrolling, by the title, maybe I expected too much 😆.

**Comment by tobbi** [comment]
Approval from me once the review comments were addressed.

**Comment by tobbi** [comment]
Oh, and check the travis warnings and fix them.

**Comment by Semphriss** [comment]
I'm gonna fix all of this, I'd just have one question :

> Btw, I expected mouse scrolling, by the title, maybe I expected too much.

I'm not sure to understand... It scrolls by hovering the first or last 32 pixels of the widget bar, isn't that the way it should work? I wanted to make it work the same way as the tilesets widget. Do you want me to make it another way? What's the method you originally understood from the title?

File: src/editor/layer_icon.cpp (hunk 1 above)
Comment (Semphriss) [reply src/editor/layer_icon.cpp:70]: Dang, I though I'd catch 'em all with a `git diff`... Gonna review the whole thing :laughing:

**Comment by Zwatotem** [comment]
Sorry, my comment didn't make much sense. What I had in mind was mouse wheel scrolling, just like the tile selection panel does, but horizontally.

**Comment by Semphriss** [comment]
Oh, okay. Should I also make it scrollable with mouse wheel?

**Comment by Zwatotem** [comment]
If it isn't much of a deal for you then sure.

**Comment by Semphriss** [comment]
I'll look into this; I'm pushing my first fix batch in a few moments, I'll use that first batch to see if there are any other Travis errors.

I'll add the mouse wheel scroll (if doable in less than an hour :laughing:) in the second next commit, along with any other Travis errors if applicable.

**Comment by Semphriss** [comment]
Alright, I couldn't resist and I did everything in a single commit.

I do have a problem, though; scrolling with the mouse wheel also triggers the editor behind the widget. It seems like my `return true;` in the mouse wheel function isn't enough to prevent propagation.

**Comment by Semphriss** [comment]
Nevermind, it's a code 18. It works fine now, and I fixed all the requested changes; I'll commit now.

File: src/editor/layers_widget.cpp [hunk 3]
```diff
@@ -108,7 +110,11 @@ EditorLayersWidget::draw(DrawingContext& context)
   int pos = 0;
   for (const auto& layer_icon : m_layer_icons) {
     if (layer_icon->is_valid()) {
-      layer_icon->draw(context, get_layer_coords(pos));
+      if (pos*35 >= m_scroll) {
+        layer_icon->draw(context, get_layer_coords(pos));
+      } else if ((pos+1)*35 >= m_scroll) {
+        layer_icon->draw(context, get_layer_coords(pos), 35 - (m_scroll - pos*35));
+      }
```
Comment (Zwatotem) [inline src/editor/layers_widget.cpp:117]: Operators still glued to operands here...

File: src/editor/object_icon.cpp [hunk 4]
```diff
@@ -65,7 +65,26 @@ void
...
+                                        pos + m_offset,
+                                        pos + Vector(32, 32) - m_offset
+                                      ),
+                                      LAYER_GUI - 9);
+}
+
+void
+ObjectIcon::draw(DrawingContext& context, const Vector& pos, int pixels_shown)
+{
+  auto cropped_surface = m_surface->region(Rect(32-pixels_shown,
```
Comment (Zwatotem) [inline src/editor/object_icon.cpp:78]: ...and here

**Review by Zwatotem — COMMENTED** [review]
There is an unintended behaviour in that new feature, the bar keeps scrolling, even when I take the mouse.
See video:
https://youtu.be/BV5S4DYmnbU

**Pull request merged**